src/routes.rs

```diff
@@ -286,21 +297,41 @@ async fn fetch_logs_from_blocks(
 
     if all_blocks_cached {
         return Ok(cached_logs);
+        // just return results and don't continue with the code below
     }
 
-    // if not all blocks are cached, go fetch
+    // sempaphore to limit number of concurrent requests
+    // let _permit = semaphore.acquire_owned().await.unwrap();
+    // println!("Acquired permit");
+
+    // if not all blocks are cached, go fetch and store to cache
     println!(
         "\nFetching logs from Infura for blocks {} to {}",
         start_block, end_block
     );
     let infura_url = env::var("INFURA_URL").expect("INFURA_URL must be set");
+
+    // // Acquire the rate limiter lock and perform the check
+    // let is_allowed = {
+    //     let rl = rate_limiter.lock().await;
+    //     rl.check().is_ok()
+    // };
+
+    // if is_allowed {
+    // Wait for the rate limiter's permission before making the API call
+    let mut rate_limiter_guard = rate_limiter.lock().await;
+    rate_limiter_guard.until_ready().await;
+    drop(rate_limiter_guard); // Release the lock
```

Comment:
You could just use a nested block to auto-drop the `rate_limiter_guard`
```suggestion
    {
      let mut rate_limiter_guard = rate_limiter.lock().await;
      rate_limiter_guard.until_ready().await;
    }
```